src/main/java/Duke.java

```diff
@@ -1,10 +1,172 @@
+import command.PrintListCommand;
+import command.task.AddDeadlineTaskCommand;
+import command.task.AddEventTaskCommand;
+import command.task.AddTodoTaskCommand;
+import java.util.ArrayList;
+import java.util.Scanner;
+import java.util.stream.Collectors;
+import parser.CommandParser;
+import parser.TaskParser;
+import store.Store;
+import task.Task;
+import ui.CommandLineUi;
+import ui.Ui;
+
+/**
+ * A task reminder command line application.
+ */
 public class Duke {
-    public static void main(String[] args) {
-        String logo = " ____        _        \n"
-                + "|  _ \\ _   _| | _____ \n"
-                + "| | | | | | | |/ / _ \\\n"
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+
+  private Scanner scanner = new Scanner(System.in);
+
+  private ArrayList<Task> list = new ArrayList<>();
+
+  private Ui ui = new CommandLineUi();
+
+  private CommandParser commandParser;
+
+  private TaskParser taskParser;
+
+  private Store store = new Store(ui::error);
+
+  /**
+   * The entry point of application.
+   *
+   * @param args (unused)
+   */
+  public static void main(String[] args) {
+    Duke duke = new Duke();
+    duke.printWelcome();
+    duke.startRepl();
+  }
+
+  public Duke(Ui ui) {
+    this(ui, new Store(ui::error));
+  }
+
+  /**
+   * Instantiates a new Duke.
+   *
+   * @param ui    the ui to output to the user
+   * @param store the store to store and retrieve tasks
+   */
+  public Duke(Ui ui, Store store) {
+    this.ui = ui;
+    this.store = store;
+    taskParser = new TaskParser(ui);
+    initalizeParser();
+  }
+
+  /**
+   * Instantiates a new Duke.
+   */
+  public Duke() {
+    taskParser = new TaskParser(ui);
+    initalizeParser();
+  }
+
+  private void initalizeParser() {
+    commandParser = new CommandParser(ui);
+    commandParser.add(new ByeCommand(ui::exit));
+    commandParser.add(new PrintListCommand(this::printList));
+    commandParser.add(new MarkAsDoneCommand(ui::error, this::findInListThenMarkAsDone));
+    commandParser.add(new DeleteCommand(ui::error, this::findInListThenDelete));
+    commandParser.add(new AddTodoTaskCommand(ui::error, this::addTaskToList));
+    commandParser.add(new AddEventTaskCommand(ui::error, this::addTaskToList));
+    commandParser.add(new AddDeadlineTaskCommand(ui::error, this::addTaskToList));
+    commandParser.add(new FindCommand(ui, this::findTasksMatchingQuery));
+    loadFromDiskToList();
+  }
+
+  private void printWelcome() {
+    String logo = " ____        _        \n"
+        + "|  _ \\ _   _| | _____ \n"
+        + "| | | | | | | |/ / _ \\\n"
+        + "| |_| | |_| |   <  __/\n"
+        + "|____/ \\__,_|_|\\_\\___|\n";
+    ui.println("Hello from\n" + logo);
+    ui.println("What can I do for you?");
+
+  }
+
+  private void startRepl() {
+    while (true) {
+      parseAndExecuteCommand(scanner.nextLine().trim());
     }
+  }
+
```

Comment:
Could probably add in JavaDocs for methods without JavaDocs, making it easier to understand